<a href="https://colab.research.google.com/github/riyajaiswal25/SVM/blob/main/creditcard.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv("creditcard.csv")

In [3]:
data

Time        V1        V2        V3        V4        V5        V6  \
0         0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1         0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2         1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3         1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4         2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...     ...       ...       ...       ...       ...       ...       ...   
7968  10980  1.284388 -0.013181  0.646174  0.198985 -0.568675 -0.526121   
7969  10981  1.190428 -0.122329  0.954945  0.267101 -0.971026 -0.652279   
7970  10981 -0.725175  0.298202  1.824761 -2.587170  0.283605 -0.016617   
7971  10981  1.226153 -0.129645  0.735197  0.142752 -0.703245 -0.349641   
7972  10981  1.145381 -0.059349  0.968088  0.267891 -0.822582 -0.597727   

            V7        V8        V9  ...       V21       V22       V23  \
0     0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1    -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2     0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3     0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4     0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...        ...       ...       ...  ...       ...       ...       ...   
7968 -0.448235 -0.167709  1.773223  ... -0.101868 -0.030298 -0.081412   
7969 -0.612992 -0.003909  1.633117  ... -0.015001  0.127027  0.012079   
7970  0.153659  0.045084 -0.197611  ... -0.017097 -0.070535 -0.442861   
7971 -0.612641  0.020507  1.648986  ... -0.047936  0.040196 -0.057391   
7972 -0.450197 -0.119747  1.338188  ...       NaN       NaN       NaN   

           V24       V25       V26       V27       V28  Amount  Class  
0     0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62    0.0  
1    -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69    0.0  
2    -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66    0.0  
3    -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50    0.0  
4     0.141267 -0.206010  0.502292  0.219422  0.215153   69.99    0.0  
...        ...       ...       ...       ...       ...     ...    ...  
7968 -0.123281  0.278808  1.064001 -0.090181  0.000481   15.95    0.0  
7969  0.534409  0.112179  1.004483 -0.100188 -0.004774   14.95    0.0  
7970 -0.895837  0.624743 -0.510601 -0.031142  0.025564   12.95    0.0  
7971 -0.012386  0.187685  1.037786 -0.100081 -0.009869   15.95    0.0  
7972       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[7973 rows x 31 columns]

In [5]:
data.describe()

Time           V1           V2           V3           V4  \
count   7973.000000  7973.000000  7973.000000  7973.000000  7973.000000   
mean    4257.151261    -0.299740     0.295226     0.899355     0.215736   
std     3198.964299     1.498341     1.283914     1.090297     1.447057   
min        0.000000   -23.066842   -25.640527   -12.389545    -4.657545   
25%     1531.000000    -1.046362    -0.237359     0.372435    -0.687521   
50%     3635.000000    -0.416341     0.335446     0.948695     0.223379   
75%     6662.000000     1.122758     0.950582     1.597949     1.131542   
max    10981.000000     1.685314     8.261750     4.101716     7.380245   

                V5           V6           V7           V8           V9  ...  \
count  7973.000000  7973.000000  7973.000000  7973.000000  7973.000000  ...   
mean     -0.025285     0.157286    -0.026445    -0.070525     0.655244  ...   
std       1.167218     1.325015     1.063709     1.332568     1.156618  ...   
min     -32.092129    -7.574798   -12.968670   -23.632502    -3.878658  ...   
25%      -0.630525    -0.655399    -0.517733    -0.199794    -0.085635  ...   
50%      -0.107337    -0.148669     0.004732     0.016128     0.613170  ...   
75%       0.405082     0.555200     0.527353     0.307111     1.294087  ...   
max      11.974269    21.393069    34.303177     3.877662    10.392889  ...   

               V21          V22          V23          V24          V25  \
count  7972.000000  7972.000000  7972.000000  7972.000000  7972.000000   
mean     -0.053715    -0.165799    -0.035174     0.025977     0.088893   
std       0.953498     0.654858     0.488322     0.601760     0.427505   
min     -11.468435    -8.527145   -15.144340    -2.512377    -2.577363   
25%      -0.271837    -0.581473    -0.182989    -0.340419    -0.161009   
50%      -0.130344    -0.167048    -0.046107     0.089606     0.115418   
75%       0.044823     0.250886     0.086806     0.421015     0.361249   
max      22.588989     4.534454    13.876221     3.200201     5.525093   

               V26          V27          V28       Amount        Class  
count  7972.000000  7972.000000  7972.000000  7972.000000  7972.000000  
mean      0.020256     0.016150     0.001161    65.413540     0.003136  
std       0.517409     0.403570     0.275976   194.911169     0.055915  
min      -1.338556    -7.976100    -3.054085     0.000000     0.000000  
25%      -0.363180    -0.063198    -0.019081     4.617500     0.000000  
50%      -0.015260     0.007101     0.018443    15.950000     0.000000  
75%       0.329322     0.144700     0.080563    54.910000     0.000000  
max       3.517346     4.173387     4.860769  7712.430000     1.000000  

[8 rows x 31 columns]

In [7]:
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7973 entries, 0 to 7972
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    7973 non-null   int64  
 1   V1      7973 non-null   float64
 2   V2      7973 non-null   float64
 3   V3      7973 non-null   float64
 4   V4      7973 non-null   float64
 5   V5      7973 non-null   float64
 6   V6      7973 non-null   float64
 7   V7      7973 non-null   float64
 8   V8      7973 non-null   float64
 9   V9      7973 non-null   float64
 10  V10     7973 non-null   float64
 11  V11     7973 non-null   float64
 12  V12     7973 non-null   float64
 13  V13     7973 non-null   float64
 14  V14     7973 non-null   float64
 15  V15     7972 non-null   float64
 16  V16     7972 non-null   float64
 17  V17     7972 non-null   float64
 18  V18     7972 non-null   float64
 19  V19     7972 non-null   float64
 20  V20     7972 non-null   float64
 21  V21     7972 non-null   float64
 22  

In [9]:
# Import necessary libraries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import classification_report, confusion_matrix

In [10]:
# Drop rows with missing values for simplicity
data = data.dropna()


In [11]:
data


Time        V1        V2        V3        V4        V5        V6  \
0         0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1         0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2         1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3         1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4         2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...     ...       ...       ...       ...       ...       ...       ...   
7967  10980 -0.046786  0.030050  2.037794 -0.670130 -0.727283 -0.588537   
7968  10980  1.284388 -0.013181  0.646174  0.198985 -0.568675 -0.526121   
7969  10981  1.190428 -0.122329  0.954945  0.267101 -0.971026 -0.652279   
7970  10981 -0.725175  0.298202  1.824761 -2.587170  0.283605 -0.016617   
7971  10981  1.226153 -0.129645  0.735197  0.142752 -0.703245 -0.349641   

            V7        V8        V9  ...       V21       V22       V23  \
0     0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1    -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2     0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3     0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4     0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...        ...       ...       ...  ...       ...       ...       ...   
7967 -0.067966 -0.370767  0.228931  ...  0.264364  1.078896 -0.097768   
7968 -0.448235 -0.167709  1.773223  ... -0.101868 -0.030298 -0.081412   
7969 -0.612992 -0.003909  1.633117  ... -0.015001  0.127027  0.012079   
7970  0.153659  0.045084 -0.197611  ... -0.017097 -0.070535 -0.442861   
7971 -0.612641  0.020507  1.648986  ... -0.047936  0.040196 -0.057391   

           V24       V25       V26       V27       V28  Amount  Class  
0     0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62    0.0  
1    -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69    0.0  
2    -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66    0.0  
3    -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50    0.0  
4     0.141267 -0.206010  0.502292  0.219422  0.215153   69.99    0.0  
...        ...       ...       ...       ...       ...     ...    ...  
7967  0.375679 -0.500253 -0.159051 -0.018267 -0.061794   39.00    0.0  
7968 -0.123281  0.278808  1.064001 -0.090181  0.000481   15.95    0.0  
7969  0.534409  0.112179  1.004483 -0.100188 -0.004774   14.95    0.0  
7970 -0.895837  0.624743 -0.510601 -0.031142  0.025564   12.95    0.0  
7971 -0.012386  0.187685  1.037786 -0.100081 -0.009869   15.95    0.0  

[7972 rows x 31 columns]

In [13]:
# Separate features (X) and target variable (y)
X = data.drop('Class', axis=1)
y = data['Class']

In [14]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [15]:
# Standardize the features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [16]:
# Create SVM model
svm_model = SVC(kernel='linear', C=1)

In [17]:
# Train the model
svm_model.fit(X_train, y_train)

SVC(C=1, kernel='linear')

In [18]:
# Make predictions on the test set
y_pred = svm_model.predict(X_test)

In [19]:
# Evaluate the model
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Confusion Matrix:
 [[1590    2]
 [   0    3]]

Classification Report:
               precision    recall  f1-score   support

         0.0       1.00      1.00      1.00      1592
         1.0       0.60      1.00      0.75         3

    accuracy                           1.00      1595
   macro avg       0.80      1.00      0.87      1595
weighted avg       1.00      1.00      1.00      1595




# Summary

**True Positives (TP)**: 3
**True Negatives (TN)**: 1590
**False Positives (FP)**: 2
**False Negatives (FN)**: 0

***Interpretation***:

**Accuracy**: The overall accuracy of the model is excellent (close to 100%), indicating that the model correctly predicted the majority of instances.

**Precision**: Precision for class 0 (non-fraud) is perfect (1.00), meaning that when the model predicts a transaction as non-fraudulent, it is almost always correct. For class 1 (fraud), precision is 0.60, indicating that when the model predicts fraud, it is correct 60% of the time.

**Recall (Sensitivity)**: Recall for class 0 is perfect (1.00), indicating that the model captures all actual non-fraudulent transactions. For class 1, recall is also perfect (1.00), suggesting that the model effectively identifies all actual fraudulent transactions.

**F1-Score**: The F1-score is the harmonic mean of precision and recall. The weighted average F1-score is close to 1.00, indicating a good balance between precision and recall.

**Support**: The number of instances in each class. Class 0 has a high support (1592), while class 1 has a lower support (3) due to the imbalanced nature of the dataset.

In summary,the SVM model appears to perform very well, especially in correctly identifying non-fraudulent transactions (class 0). However, due to the imbalanced nature of the dataset, where fraudulent transactions are rare, the model's precision for fraud detection (class 1) is slightly lower.
